### PR TITLE
Upgrade to Ember CLI 0.2.0 and Ember Data 1.0.0-beta.15

### PR DESCRIPTION
**Pull request by notmessenger**
Incorporates the following commits from joshfester/ember-parse-adapter:

337d2ff
6cf0edc
fc2618f
71afa20
139eb5f
c486693
a7aaede
de72f1b
121909d
e79b4f0
8074563

### CODE REVIEW

File: addon/serializers/application.js [hunk 1]
```diff
@@ -86,9 +86,18 @@ export default DS.RESTSerializer.extend({
...
+          // this is an object that make ember-data fail, as it expects nothing or an array ids that represent the records
+          hash[key] = [];
+
+          // ember-data expects the link to be a string
+          // The adapter findHasMany will parse it
+          if (!hash.links) {
+            hash.links = {};
+          }
+
+          hash.links[key] = JSON.stringify({typeKey: relationship.type.typeKey, key: key});
```
Comment (mixonic) [inline addon/serializers/application.js]: `typeKey` as the key? Not `type`?

File: addon/serializers/application.js [hunk 2]
```diff
@@ -190,8 +190,8 @@ export default DS.RESTSerializer.extend({
       hasMany.forEach( function( child ) {
         json[key].objects.push({
           '__type'    : 'Pointer',
-          'className' : child.parseClassName(),
-          'objectId'  : child.get( 'id' )
+          'className' : _this.parseClassName(child.type.typeKey),
```
Comment (mixonic) [inline addon/serializers/application.js]: In `serializeBelongsTo` the `key` (used in serialization to the server) is passed to `parseClassName`. Here however the `typeKey` is passed, the Ember-Data internal value.

IMO only the `typeKey` should be passed to `parseClassName`, but can you clarify?

File: addon/serializers/application.js (hunk 1 above)
Comment (notmessenger) [inline addon/serializers/application.js]: https://github.com/joshfester/ember-parse-adapter/commit/121909ddcc417f580b931387888c7ce5d75d722a introduced the JSON.stringify() call and https://github.com/joshfester/ember-parse-adapter/commit/337d2ff8fc4c3d6afeccf65145e71ca025623f99 changed "type" to "typeKey"

I mean to include the commit messages from each of these hashes into this commit listed under each referenced hash.  If this is preferred I can cancel this pull request and re-commit.

File: addon/serializers/application.js (hunk 2 above)
Comment (notmessenger) [inline addon/serializers/application.js]: This is something hopefully @joshfester or @Kira2 can speak to.  I'm honestly just going through the commits at github.com/joshfester/ember-parse-adapter and extracting the ones that are applicable to Ember Data 1.0.0-beta.15 and replaying them onto the Ember CLI Add-on version of the adapter.  I will then do the same thing with other commits against Ember Data 1.0.0-beta.16 and again with other generic fixes, etc.

I have not been focusing on the specific changes that were applied or why - if the tests are passing and the demo app still functions I have viewed it as being functionally correct.  I am trying to facilitate the convergence of several different development efforts into this single repository in the Ember CLI format.  At that point I, and others, can begin focusing specifically on refactors and improvements.

If there is a different approach you would rather I and others to follow I am open to suggestions - I am just trying to move progress forward.

File: addon/serializers/application.js (hunk 1 above)
Comment (joshfester) [inline addon/serializers/application.js]: These changes were from a pull request by @Kira2 so hopefully we can get her involved. Without running any tests I'm not sure of the difference between typeKey and type. @mixonic could you elaborate on that?

**Comment by Kira2** [comment]
Hi,

In both serializeBelongsTo and serializeBelongsTo, I use the type key.

```javascript
// into serializeBelongsTo
json[key] = {
  "__type": "Pointer",
  "className": this.parseClassName(belongsTo.typeKey),
  "objectId": belongsToId
};
```

```javascript
// into serializeBelongsTo
var addOperation, removeOperation;
var parseClassName = this.parseClassName(relationship.type.typeKey);
```

To be honest, all the changes I have made was done empirically, while I was trying to use this adapter to communicate with Parse for one of my project. I am still working on it, but I have some deadlines to respect for my other projects, so I have not found time to write enhanced tests yet. This is also why I did not created a pull request on my last changes to @joshfester repository. I did not want to do this without having stronger tests on it.

But if you want, I can create a pull request on your repository to let you see the last changes I have made since, trying to fix some more bugs. You can also see them on the "beta" branch of my repository, here is a list of them:

* fixed requestPasswordReset
* avoid "400 bad request error" when try to save an object with a null belongTo relation
* fixed belongsTo normalization when the request was made with the "include" parameter
* fixed normalization of included objects returned by Parse

The "master" branch contains some more code that is very, very experimental:

* track objects removed from a relation
* do not automatically load an entire relation when do a "get" on it, if I don't want to. I made this, because if you call "reload" on a relation with ember-data, and that relation was never loaded before, it will load all the data twice from the server.

Hope it might help.

File: addon/serializers/application.js (hunk 1 above)
Comment (mixonic) [inline addon/serializers/application.js]: `typeKey` is an Ember-Data internal string version of the model name. Creating a payload for push that looks like:

```
{
  typeKey: 'someFoo'
}
```

Would not actually set the `typeKey`. Instead polymorphic payload are intended to pass their type in a payload as `type`:

```
{
  type: 'someFoo'
}
```

In short, pushing `typeKey:` into the store doesn't so anything afaik.

File: addon/serializers/application.js (hunk 2 above)
Comment (mixonic) [inline addon/serializers/application.js]: The `serializeBelongsTo` in this code does not appear to be the version @Kira2 was talking about.

**Comment by QuantumInformation** [comment]
Keep up the good work, really hope to use this in prod soon!

**Pull request merged**